spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,321 @@
+
+    ExecutorService commiterService = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setNameFormat("Committer-Service").build());
+
+    commiterService.execute(() -> {
+      try {
+        while (stillRewriting.get() || completedRewrite.size() > groupsPerCommit) {
+          Thread.yield();
+          if (completedRewrite.size() > groupsPerCommit) {
+            // Gather a group of rewrites to do a commit
+            Set<String> batch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+            for (int i = 0; i < groupsPerCommit; i++) {
+              batch.add(completedRewrite.poll());
+            }
+            // Running commit on the group we gathered
+            try {
+              commitFileGroups(batch);
+            } catch (Exception e) {
+              if (!partialProgressEnabled) {
+                LOG.error("Failure during rewrite commit process, partial progress not enabled. Rethrowing", e);
+                throw e;
+              } else {
+                LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+              }
+            }
+            completedCommits.addAll(batch);
+          }
+        }
+        Set<String> finalBatch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+        while (completedRewrite.size() != 0) {
+          finalBatch.add(completedRewrite.poll());
+        }
+        if (finalBatch.size() != 0) {
+          // Final Commit
+          commitFileGroups(finalBatch);
+          completedCommits.addAll(finalBatch);
+        }
+      } catch (Exception e) {
+        if (!partialProgressEnabled) {
+          LOG.error("Cannot commit rewrite and partial progress not enabled, removing all completed work", e);
+          completedCommits.forEach(this::abortFileGroup);
+        }
+      }
+    });
+
+    // Start rewrite tasks
+    Tasks.Builder<Pair<FileGroupInfo, List<FileScanTask>>> rewriteTaskBuilder = Tasks.foreach(jobStream.iterator())
+        .suppressFailureWhenFinished()
+        .executeWith(rewriteService)
+        .noRetry()
+        .onFailure((info, exception) -> {
+          LOG.error("Failure during rewrite process for group {}", info.first(), exception);
+          abortFileGroup(info.first().setId);
+        });
+
+    if (partialProgressEnabled) {
+      rewriteTaskBuilder = rewriteTaskBuilder.suppressFailureWhenFinished();
+    } else {
+      rewriteTaskBuilder = rewriteTaskBuilder.stopOnFailure();
+    }
+
+    rewriteTaskBuilder
+        .run(i -> {
+          String setId = i.first().setId();
+          Set<DataFile> addedFiles = strategy.rewriteFiles(setId, i.second());
+          completedRewrite.offer(setId);
+          results.put(setId, Pair.of(i.first(), new FileGroupRewriteResult(addedFiles.size(), i.second().size())));
+        });
+    stillRewriting.set(false);
+    commiterService.shutdown();
+
+    try {
+      commiterService.awaitTermination(10, TimeUnit.MINUTES);
```

Comment:
Does it mean we give 10 minutes for the committer thread to gracefully shut down?

Reply:
Yep, this is the amount of time we give the service after all rewrites have completed (Tasks.run blocks) for all commits to finish.

So basically once all rewrite jobs have completed the remaining commits have 10 minutes to execute